aws-distro-opentelemetry-node-autoinstrumentation/src/patches/aws/services/bedrock.ts

```diff
@@ -0,0 +1,207 @@
+import { AwsSpanProcessingUtil } from '../../../aws-span-processing-util';
+
+const AGENT_ID: string = 'agentId';
+const KNOWLEDGE_BASE_ID: string = 'knowledgeBaseId';
+const DATA_SOURCE_ID: string = 'dataSourceId';
+const GUARDRAIL_ID: string = 'guardrailId';
+const MODEL_ID: string = 'modelId';
+const AWS_BEDROCK_SYSTEM: string = 'aws_bedrock';
+
+const AGENT_OPERATIONS = [
+  'CreateAgentActionGroup',
+  'CreateAgentAlias',
+  'DeleteAgentActionGroup',
+  'DeleteAgentAlias',
+  'DeleteAgent',
+  'DeleteAgentVersion',
+  'GetAgentActionGroup',
+  'GetAgentAlias',
+  'GetAgent',
+  'GetAgentVersion',
+  'ListAgentActionGroups',
+  'ListAgentAliases',
+  'ListAgentKnowledgeBases',
+  'ListAgentVersions',
+  'PrepareAgent',
+  'UpdateAgentActionGroup',
+  'UpdateAgentAlias',
+  'UpdateAgent',
+];
+
+const KNOWLEDGE_BASE_OPERATIONS = [
+  'AssociateAgentKnowledgeBase',
+  'CreateDataSource',
+  'DeleteKnowledgeBase',
+  'DisassociateAgentKnowledgeBase',
+  'GetAgentKnowledgeBase',
+  'GetKnowledgeBase',
+  'ListDataSources',
+  'UpdateAgentKnowledgeBase',
+];
+
+const DATA_SOURCE_OPERATIONS = ['DeleteDataSource', 'GetDataSource', 'UpdateDataSource'];
+
+const operationToBedrockAgentServiceMapping: { [key: string]: { [key: string]: string } } = {};
+for (const operation of AGENT_OPERATIONS) {
+  operationToBedrockAgentServiceMapping[operation] = { [AWS_ATTRIBUTE_KEYS.AWS_BEDROCK_AGENT_ID]: AGENT_ID };
+}
+for (const operation of KNOWLEDGE_BASE_OPERATIONS) {
+  operationToBedrockAgentServiceMapping[operation] = {
+    [AWS_ATTRIBUTE_KEYS.AWS_BEDROCK_KNOWLEDGE_BASE_ID]: KNOWLEDGE_BASE_ID,
+  };
+}
+for (const operation of DATA_SOURCE_OPERATIONS) {
+  operationToBedrockAgentServiceMapping[operation] = {
+    [AWS_ATTRIBUTE_KEYS.AWS_BEDROCK_DATA_SOURCE_ID]: DATA_SOURCE_ID,
+  };
+}
+
+export class BedrockAgentServiceExtension implements ServiceExtension {
+  bedrockAgentServiceInfo: { [key: string]: string } = {};
+  requestPreSpanHook(
+    request: NormalizedRequest,
+    config: AwsSdkInstrumentationConfig,
+    diag: DiagLogger
+  ): RequestMetadata {
+    const spanAttributes: Attributes = {};
+    const isIncoming = false;
+    const spanKind: SpanKind = SpanKind.CLIENT;
+    let spanName: string | undefined;
+
+    const operation: string = request.commandName;
+    if (operation && Object.keys(operationToBedrockAgentServiceMapping).includes(operation)) {
+      this.bedrockAgentServiceInfo = operationToBedrockAgentServiceMapping[operation];
+      // This should only trigger once; it is the easy way to access the 0th and only entry
+      for (const serviceInfo of Object.entries(this.bedrockAgentServiceInfo)) {
+        const [attributeKey, requestParamKey] = serviceInfo;
+        const requestParamValue = request.commandInput?.[requestParamKey];
+
+        if (requestParamValue) {
+          spanAttributes[attributeKey] = requestParamValue;
```

Comment:
The loop only runs once at most since we only ever return an object with one entry in it. I can make the improvement, but was following the Python implementation closely. Small change though so lmk if you think there's still a reason to add it here

https://github.com/aws-observability/aws-otel-python-instrumentation/blob/main/aws-opentelemetry-distro/src/amazon/opentelemetry/distro/patches/_bedrock_patches.py#L171